src/array-map.ts

```diff
@@ -0,0 +1,293 @@
+ *
+ *  class MyClass {
+ *    wrappedRecords = map({
+ *      data: () => this.records,
+ *      map: (record) => new SomeWrapper(record),
+ *    }),
+ *  }
+ * ```
+ */
+export function arrayMap<
+  Elements extends readonly unknown[],
+  MapTo = unknown,
+>(options: {
+  /**
+   * Array of non-primitives to map over
+   *
+   * This can be class instances, plain objects, or anything supported by WeakMap's key
+   */
+  data: () => Elements;
+  /**
+   * How to transform each element from `data`,
+   * similar to if you were to use Array map yourself.
+   *
+   * This function will be called only when needed / on-demand / lazily.
+   * - if iterating over part of the data, map will only be called for the elements observed
+   * - if not iterating, map will only be called for the elements observed.
+   */
+  map: (element: Elements[0]) => MapTo;
+}): MappedArray<Elements, MapTo> {
+  let { data, map } = options;
+
+  return new TrackedArrayMap(data, map) as MappedArray<Elements, MapTo>;
+}
+
+const AT = Symbol.for("__AT__");
+
+/**
+ * @private
+ */
+export class TrackedArrayMap<Element = unknown, MappedTo = unknown>
+  implements MappedArray<Element[], MappedTo>
+{
+  // Tells TS that we can array-index-access
+  [index: number]: MappedTo;
+
+  // these can't be real private fields
+  // until @cached is a real decorator
+  private _mapCache = new WeakMap<Element & object, MappedTo>();
+  private _dataFn: () => readonly Element[];
+  private _mapper: (element: Element) => MappedTo;
+
+  constructor(
+    data: () => readonly Element[],
+    map: (element: Element) => MappedTo,
+  ) {
+    this._dataFn = data;
+    this._mapper = map;
+
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const self = this;
+
+    /**
+     * This is what allows square-bracket index-access to work.
+     *
+     * Unfortunately this means the returned value is
+     * Proxy -> Proxy -> wrapper object -> *then* the class instance
+     *
+     * Maybe JS has a way to implement array-index access, but I don't know how
+     */
+    return new Proxy(this, {
+      get(_target, property) {
+        if (typeof property === "string") {
+          let parsed = parseInt(property, 10);
+
+          if (!isNaN(parsed)) {
+            return self[AT](parsed);
+          }
+        }
+
+        return self[property as keyof MappedArray<Element[], MappedTo>];
+      },
+      // Is there a way to do this without lying to TypeScript?
+    }) as TrackedArrayMap<Element, MappedTo>;
+  }
+
+  @signal
+  get _records(): (Element & object)[] {
+    let data = this._dataFn();
+
+    if (!data.every((datum) => typeof datum === "object")) {
+      throw new Error(
+        `Every entry in the data passed to \`map\` must be an object.`,
+      );
+    }
+
+    return data as Array<Element & object>;
+  }
+
+  values = () => [...this];
+
+  get length() {
+    return this._records.length;
+  }
+
+  [Symbol.iterator](): Iterator<MappedTo> {
```

Comment:
Is this behaviour-preserving from `Array.prototype[@@iterator]` if array is mutated while the iterator is alive? Is that something we want to catch and explicitly error about, like below in `[AT]`?

Reply:
I hadn't thought of mutating _while_ iterating -- sounds dangerous.
This isn't inherently doing anything to copy whatever the default behavior is.